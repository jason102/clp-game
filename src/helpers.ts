export const GAME_TIME = 5000; // 5 seconds

export const CHART_TIME_VALUES = Array.from(
  Array(GAME_TIME / 500).keys()
).reduce(
  (accum) => (accum.length ? [...accum, accum[accum.length - 1] + 0.5] : [0.5]),
  [] as number[]
);

const getNumberOfClicksInRange = (
  rangeStart: number,
  rangeEnd: number,
  clickTimestamps: number[]
) => {
  return clickTimestamps.reduce((count, nextTimestamp) => {
    if (nextTimestamp > rangeStart && nextTimestamp <= rangeEnd) {
      return count + 1;
    }
    return count;
  }, 0);
};

export const getClicksPerHalfSecond = (
  startingTime: number,
  timestamps: number[]
) => {
  let totalFoundClickCount = 0;

  const clickMap = CHART_TIME_VALUES.reduce((accumulated, next) => {
    const rangeStart = startingTime + (next - 0.5) * 1000;
    const rangeEnd = rangeStart + 500;

    const clickCount = getNumberOfClicksInRange(
      rangeStart,
      rangeEnd,
      timestamps
    );

    accumulated[next] = clickCount;
    totalFoundClickCount += clickCount;

    return accumulated;
  }, {} as { [key: number]: number });

  // Sometimes the very first click might fall slightly outside of the range and therefore we handle it here
  if (totalFoundClickCount < timestamps.length) {
    clickMap[0.5] += 1;
  }

  return Object.values(clickMap);
};

export const noLineDataArray = Array(CHART_TIME_VALUES.length).fill(0);
