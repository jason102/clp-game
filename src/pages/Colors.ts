export const ORANGE = '#ff9559';
export const BLUE = '#007aff';
