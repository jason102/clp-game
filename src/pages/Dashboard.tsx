import React, { useEffect, useRef } from 'react';
import * as echarts from 'echarts/core';
import { LineChart } from 'echarts/charts';
import {
  TitleComponent,
  GridComponent,
  DatasetComponent,
  TooltipComponent,
} from 'echarts/components';
import { CanvasRenderer } from 'echarts/renderers';
import { Container } from './Dashboard.styles';

echarts.use([
  TitleComponent,
  GridComponent,
  DatasetComponent,
  CanvasRenderer,
  TooltipComponent,
  LineChart,
]);

const Dashboard: React.FC = () => {
  const lineChartRef = useRef<any>(null);

  useEffect(() => {
    if (lineChartRef.current) {
      const lineChart = echarts.init(lineChartRef.current);

      lineChart.setOption({
        grid: { top: 8, right: 8, bottom: 24, left: 36 },
        xAxis: {
          type: 'category',
          data: ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'],
        },
        yAxis: {
          type: 'value',
        },
        series: [
          {
            data: [820, 932, 901, 934, 1290, 1330, 1320],
            type: 'line',
            smooth: true,
          },
        ],
        tooltip: {
          trigger: 'axis',
        },
      });
    }
  }, []);

  return (
    <Container>
      <div style={{ width: '60vw', height: '400px' }} ref={lineChartRef} />
    </Container>
  );
};

export default Dashboard;
