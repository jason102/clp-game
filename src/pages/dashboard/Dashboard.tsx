import React, { useEffect, useState } from 'react';
import Chart from './Chart';
import {
  Container,
  OrangeButtonCount,
  BlueButtonCount,
  ButtonCountContainer,
} from './Dashboard.styles';
import useInitDashboard from './useInitDashboard';
import { getClicksPerHalfSecond, noLineDataArray } from 'helpers';

const Dashboard: React.FC = () => {
  const [blackLineValues, setBlackLineValues] =
    useState<number[]>(noLineDataArray);
  const [orangeLineValues, setOrangeLineValues] =
    useState<number[]>(noLineDataArray);
  const [blueLineValues, setBlueLineValues] =
    useState<number[]>(noLineDataArray);
  const {
    isLoading,
    loadingError,
    orangeClickTimestamps,
    blueClickTimestamps,
    isGameFinished,
    startingTime,
  } = useInitDashboard();

  useEffect(() => {
    if (isGameFinished) {
      // Loop through all clicks and see how many were clicked per half second
      const blackLineValues = getClicksPerHalfSecond(startingTime.current, [
        ...orangeClickTimestamps,
        ...blueClickTimestamps,
      ]);

      const orangeLineValues = getClicksPerHalfSecond(
        startingTime.current,
        orangeClickTimestamps
      );

      const blueLineValues = getClicksPerHalfSecond(
        startingTime.current,
        blueClickTimestamps
      );

      setBlackLineValues(blackLineValues);
      setOrangeLineValues(orangeLineValues);
      setBlueLineValues(blueLineValues);
    }
  }, [
    blueClickTimestamps,
    isGameFinished,
    orangeClickTimestamps,
    startingTime,
  ]);

  if (isLoading) {
    return <Container>Loading Dashboard</Container>;
  }

  if (loadingError) {
    return <Container>{'Error loading dashboard: ' + loadingError}</Container>;
  }

  return (
    <Container>
      <Chart
        blackLineValues={blackLineValues}
        orangeLineValues={orangeLineValues}
        blueLineValues={blueLineValues}
      />
      <ButtonCountContainer>
        <OrangeButtonCount>{orangeClickTimestamps.length}</OrangeButtonCount>
        <BlueButtonCount>{blueClickTimestamps.length}</BlueButtonCount>
      </ButtonCountContainer>
    </Container>
  );
};

export default Dashboard;
