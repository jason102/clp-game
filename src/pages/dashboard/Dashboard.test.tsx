import { render, waitFor, act, screen } from '@testing-library/react';
import Dashboard from './Dashboard';
import * as firebase from 'firebase/firestore';
import * as helpers from 'helpers';
import { BLUE, ORANGE } from 'Colors';

jest.mock('firebaseConfig', () => ({
  blueClicksDocRef: 'blueClicksDocRef',
  orangeClicksDocRef: 'orangeClicksDocRef',
}));
jest.mock('firebase/firestore');

const echartsSpy = jest.fn();
jest.mock('echarts/core', () => ({
  use: jest.fn(),
  init: () => ({
    setOption: echartsSpy,
  }),
}));

jest.mock('echarts/charts', () => ({}));
jest.mock('echarts/components', () => ({}));
jest.mock('echarts/features', () => ({}));
jest.mock('echarts/renderers', () => ({}));

describe('Dashboard page', () => {
  beforeEach(() => {
    jest.resetAllMocks();

    jest.spyOn(firebase, 'writeBatch').mockImplementation(() => ({
      update: jest.fn(),
      set: jest.fn(),
      commit: jest.fn(),
      delete: jest.fn(),
    }));
  });

  it('renders the loading state and dashboard when loading is successful', async () => {
    const { asFragment } = render(<Dashboard />);

    // waitFor is needed as there's a promise that needs resolving in the first useEffect (Promise.all(...))
    await waitFor(() => {
      // Loading state snapshot
      expect(asFragment()).toMatchInlineSnapshot(`
        <DocumentFragment>
          <div
            class="sc-bczRLJ fptUTC"
          >
            Loading Dashboard
          </div>
        </DocumentFragment>
      `);
    });

    // Finished loading state snapshot
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders an error when resetting the game fails', async () => {
    jest.spyOn(firebase, 'writeBatch').mockImplementation(() => ({
      update: jest.fn(),
      set: jest.fn(),
      commit: () => {
        throw 'firebase string-only error yikes';
      },
      delete: jest.fn(),
    }));

    const { asFragment } = render(<Dashboard />);

    await waitFor(() => {
      expect(asFragment()).toMatchInlineSnapshot(`
        <DocumentFragment>
          <div
            class="sc-bczRLJ fptUTC"
          >
            Error loading dashboard: firebase string-only error yikes
          </div>
        </DocumentFragment>
      `);
    });
  });

  it('should show updated count values when onSnapshot fires', async () => {
    jest
      .spyOn(firebase, 'onSnapshot')
      .mockImplementationOnce((_, onNext) => {
        // @ts-ignore:
        onNext({ data: () => ({ clicks: 1 }) }); // blue clicks

        return jest.fn();
      })
      .mockImplementationOnce((_, onNext) => {
        // @ts-ignore:
        onNext({ data: () => ({ clicks: 2 }) }); // orange clicks

        return jest.fn();
      });

    render(<Dashboard />);

    await waitFor(() => {
      const blueClickCount = screen.getByText(1);
      const orangeClickCount = screen.getByText(2);

      expect(orangeClickCount).toBeInTheDocument();
      expect(orangeClickCount).toHaveStyle(`background-color: ${ORANGE}`);
      expect(blueClickCount).toBeInTheDocument();
      expect(blueClickCount).toHaveStyle(`background-color: ${BLUE}`);
    });
  });

  it('should call echarts setOption() to draw the graph after 5 seconds with click data', async () => {
    jest
      .spyOn(firebase, 'onSnapshot')
      .mockImplementationOnce((_, onNext) => {
        // @ts-ignore:
        onNext({ data: () => ({ clicks: 1 }) }); // blue clicks

        return jest.fn();
      })
      .mockImplementationOnce((_, onNext) => {
        setTimeout(() => {
          // @ts-ignore:
          onNext({ data: () => ({ clicks: 1 }) }); // orange clicks
        }, 3000); // fire an orange click after 3 seconds

        return jest.fn();
      });

    const blackClicks = [0, 0, 0, 0, 0, 1, 0, 0, 1, 0];
    const orangeClicks = [0, 0, 0, 0, 0, 0, 0, 0, 1, 0];
    const blueClicks = [0, 0, 0, 0, 0, 1, 0, 0, 0, 0];
    jest
      .spyOn(helpers, 'getClicksPerHalfSecond')
      .mockImplementationOnce(() => blackClicks)
      .mockImplementationOnce(() => orangeClicks)
      .mockImplementationOnce(() => blueClicks);

    jest.useFakeTimers();

    render(<Dashboard />);

    // Need to use act() due to multiple renders being triggered
    await act(async () => {
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      await waitFor(() => {}); // pass the await batchOperation.commit()

      jest.advanceTimersByTime(9001); // end the game after 5 seconds
    });

    expect(echartsSpy).toHaveBeenCalledWith({
      grid: { containLabel: true, left: '3%', right: '13%' },
      legend: {},
      series: [
        {
          color: 'black',
          data: blackClicks,
          name: 'Total Clicks',
          smooth: true,
          type: 'line',
        },
        {
          color: '#ff9559',
          data: orangeClicks,
          name: 'Orange',
          smooth: true,
          type: 'line',
        },
        {
          color: '#007aff',
          data: blueClicks,
          name: 'Blue',
          smooth: true,
          type: 'line',
        },
      ],
      tooltip: { trigger: 'axis' },
      xAxis: {
        axisLabel: { showMaxLabel: true, showMinLabel: true },
        boundaryGap: true,
        data: [0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5],
        name: 'Seconds',
        type: 'category',
      },
      yAxis: { name: 'Clicks', type: 'value' },
    });

    jest.useRealTimers();
  });
});
