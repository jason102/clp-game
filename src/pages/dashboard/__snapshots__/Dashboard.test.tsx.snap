// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard page renders the loading state and dashboard when loading is successful 2`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ fptUTC"
  >
    <div
      class="sc-gsnTZi iTsiio"
    />
    <div
      class="sc-dkzDqf kpvutt"
    >
      <div
        class="sc-hKMtZM sc-eCYdqJ dLVDKf ebCuMD"
      >
        0
      </div>
      <div
        class="sc-hKMtZM sc-jSMfEi dLVDKf hbJYNr"
      >
        0
      </div>
    </div>
    <p>
      To play the game, you must 
      <a
        href="/client"
        target="_blank"
      >
        open this page
      </a>
    </p>
  </div>
</DocumentFragment>
`;
